### PR TITLE
Add hidpi support to swaybar

### PR DESCRIPTION
**Pull request by SirCmpwn**
Fixes https://github.com/swaywm/sway/issues/1137

Ref https://github.com/swaywm/sway/issues/797

### CODE REVIEW

File: swaybar/bar.c [hunk 1]
```diff
@@ -262,9 +305,20 @@ void bar_setup(struct swaybar *bar,
...
-	assert(pointer->cursor_theme = wl_cursor_theme_load(NULL, 16, bar->shm));
+	int max_scale = 1;
+	struct swaybar_output *output;
+	wl_list_for_each(output, &bar->outputs, link) {
+		if (output->scale > max_scale) {
+			max_scale = output->scale;
+		}
+	}
+
+	assert(pointer->cursor_theme = wl_cursor_theme_load(
```
Comment (ascent12) [inline swaybar/bar.c]: You shouldn't put a side effect in an assertion.

Comment (SirCmpwn) [reply swaybar/bar.c]: I do this all the time, it's all over sway et al. This is fine.

Comment (SirCmpwn) [reply swaybar/bar.c]: Nevermind duh I see why

Comment (SirCmpwn) [reply swaybar/bar.c]: Pushed fix to this file but will file a follow-up ticket to go fix my fuck-ups elsewhere

**Comment by SirCmpwn** [comment]
>All of this code works but is hard to read. Converting between surface-local and buffer-local coordinates at the beginning and the end of functions doesn't help. I'd prefer everything to take coordinates in surface-local coordinates (especially hotspots), and mark variables using buffer-local coordinates with a buffer_ prefix.

Part of the problem is that we need to communciate the scale to pango so it can up the font size/etc accordingly. I think that the approach used here is reasonably good considering this constraint.

File: swaybar/render.c [hunk 2]
```diff
@@ -260,15 +290,23 @@ static uint32_t render_status_line(cairo_t *cairo,
...
+
+	int ws_vertical_padding = WS_VERTICAL_PADDING * output->scale;
+	int ws_horizontal_padding = WS_HORIZONTAL_PADDING * output->scale;
+	int border_width = BORDER_WIDTH * output->scale;
+
 	uint32_t ideal_height = text_height + ws_vertical_padding * 2
 		+ border_width * 2;
-	if (height < ideal_height) {
-		return ideal_height;
+	if (height < ideal_height / output->scale) {
```
Comment (emersion) [inline swaybar/render.c]: Here's an example of why this PR is unreadable: here you compare `height` (in buffer coordinates) with `ideal_height / output->scale` (in surface coordinates). That doesn't make sense.

File: swaybar/bar.c [hunk 3]
```diff
@@ -256,33 +299,45 @@ void bar_setup(struct swaybar *bar,
...
+	int max_scale = 1;
+	struct swaybar_output *output;
+	wl_list_for_each(output, &bar->outputs, link) {
+		if (output->scale > max_scale) {
+			max_scale = output->scale;
+		}
+	}
+
+	pointer->cursor_theme = wl_cursor_theme_load(
+				NULL, 16 * (max_scale * 2), bar->shm);
```
Comment (emersion) [inline swaybar/bar.c]: Cursor size changes when hovering the bar. Should be `24 * max_scale`.

Comment (emersion) [reply swaybar/bar.c]: What if the output scale is dynamically changed?

File: swaybar/render.c [hunk 4]
```diff
@@ -14,55 +14,73 @@
...
-	return ideal_height;
+	return ideal_height / output->scale;
 }
 
 static uint32_t render_status_line_text(cairo_t *cairo,
-		struct swaybar_config *config, const char *text,
-		bool focused, double *x, uint32_t width, uint32_t height) {
+		struct swaybar_output *output, struct swaybar_config *config,
+		const char *text, bool focused, double *x,
+		uint32_t width, uint32_t surface_height) {
```
Comment (emersion) [inline swaybar/render.c]: `surface_width`

Comment (emersion) [reply swaybar/render.c]: It's unused actually, does it make sense to remove it?

File: swaybar/render.c [hunk 5]
```diff
@@ -133,10 +157,10 @@ static uint32_t render_status_block(cairo_t *cairo,
 		if (config->sep_symbol) {
 			int _height;
 			get_text_size(cairo, config->font, &sep_width, &_height,
-					1, false, "%s", config->sep_symbol);
+					output->scale, false, "%s", config->sep_symbol);
 			uint32_t _ideal_height = _height + ws_vertical_padding * 2;
-			if (height < _ideal_height) {
-				return _height;
+			if (height < _ideal_height / output->scale) {
```
Comment (emersion) [inline swaybar/render.c]: Should that be `_height`?

File: swaybar/render.c [hunk 6]
```diff
@@ -242,33 +272,41 @@ static uint32_t render_status_line_i3bar(cairo_t *cairo,
 static uint32_t render_status_line(cairo_t *cairo,
 		struct swaybar_config *config, struct swaybar_output *output,
 		struct status_line *status, bool focused,
-		double *x, uint32_t width, uint32_t height) {
+		double *x, uint32_t width, uint32_t surface_height) {
```
Comment (emersion) [inline swaybar/render.c]: Ditto: `surface_width`?

File: swaybar/render.c [hunk 7]
```diff
@@ -330,14 +368,22 @@ static uint32_t render_workspace_button(cairo_t *cairo,
 		box_colors = config->colors.inactive_workspace;
 	}
 
+	uint32_t height = surface_height *output->scale;
```
Comment (emersion) [inline swaybar/render.c]: Oh my, missing a space

File: swaybar/render.c [hunk 8]
```diff
@@ -462,9 +510,11 @@ void render_frame(struct swaybar *bar, struct swaybar_output *output) {
...
 		cairo_paint(shm);
 
+		wl_surface_set_buffer_scale(output->surface, output->scale);
 		wl_surface_attach(output->surface,
 				output->current_buffer->buffer, 0, 0);
-		wl_surface_damage(output->surface, 0, 0, output->width, output->height);
+		wl_surface_damage(output->surface, 0, 0,
+				output->width, output->height);
 		wl_surface_commit(output->surface);
 		wl_display_roundtrip(bar->display);
```
Comment (emersion) [inline swaybar/render.c]: I still don't get why this roundtrip is needed?

File: swaybar/bar.c (hunk 3 above)
Comment (SirCmpwn) [reply swaybar/bar.c]: Aha

Comment (SirCmpwn) [reply swaybar/bar.c]: If the output scale is dynamically changed we're SOL, I want to file a ticket and deal with it never

File: swaybar/render.c (hunk 4 above)
Comment (SirCmpwn) [reply swaybar/render.c]: yup

File: swaybar/render.c (hunk 7 above)
Comment (SirCmpwn) [reply swaybar/render.c]: oh my

File: swaybar/render.c (hunk 8 above)
Comment (SirCmpwn) [reply swaybar/render.c]: We don't dispatch in a loop like some of our other clients do.

Comment (emersion) [reply swaybar/render.c]: Oh, I see, makes sense

**Pull request merged**